Sail/src/Sail/game/PlayerController.cpp

```diff
@@ -0,0 +1,150 @@
+	float speedModifier = 1.f;
+	float movementSpeed = 0.f;
+	float lookSensitivityMouse = 0.1f;
+	float lookSensitivityController = 90.0f * 0.016f;
+
+	float forwardM = 0.f, backM = 0.f, leftM = 0.f, rightM = 0.f, upM = 0.f, downM = 0.f;
+
+	MovementComponent* playerMovComp = m_player->getComponent<MovementComponent>();
+
+	// Increase speed if shift or right trigger is pressed
+	if ( Input::IsKeyPressed(SAIL_KEY_SHIFT) ) {
+		speedModifier = 5.f;
+	}
+
+	//
+	// Forwards / backwards motion
+	//
+
+	// Keyboard
+	if ( Input::IsKeyPressed(SAIL_KEY_W) ) {
+		playerMovComp->setSpeed(m_movementSpeed * speedModifier);
+		forwardM = 1.0f;
+	}
+
+	if ( Input::IsKeyPressed(SAIL_KEY_S) ) {
+		playerMovComp->setSpeed(m_movementSpeed * speedModifier);
+		backM = 1.0f;
+	}
+
+	//
+	// Side to side motion
+	//
+
+	glm::vec3 right = glm::cross(glm::vec3(0.f, 1.f, 0.f), m_cam->getCameraDirection());
+	right = glm::normalize(right);
+	// Gamepad
+	//setCameraPosition(getCameraPosition() - right * gpState.thumbSticks.leftX * movementSpeed);
+
+	// Keyboard
+	if ( Input::IsKeyPressed(SAIL_KEY_A) ) {
+		playerMovComp->setSpeed(m_movementSpeed * speedModifier);
+		leftM = 1.0f;
+	}
+	if ( Input::IsKeyPressed(SAIL_KEY_D) ) {
+		playerMovComp->setSpeed(m_movementSpeed * speedModifier);
+		rightM = 1.0f;
+	}
+
+	//
+	// Up and down motion
+	//
+
+	// Keyboard
+	if ( Input::IsKeyPressed(SAIL_KEY_SPACE) ) {
+		playerMovComp->setSpeed(m_movementSpeed * speedModifier);
+		upM = 1.0f;
+	}
+	if ( Input::IsKeyPressed(SAIL_KEY_CONTROL) ) {
+		playerMovComp->setSpeed(m_movementSpeed * speedModifier);
+		downM = 1.0f;
+	}
+
+	//
+	// Look around motion
+	//
+
+	// Mouse input
+
+	// Toggle cursor capture on right click
+	if ( Input::WasMouseButtonJustPressed(SAIL_MOUSE_RIGHT_BUTTON) ) {
+		Input::HideCursor(!Input::IsCursorHidden());
+	}
+
+	if ( Input::IsCursorHidden() ) {
+		glm::ivec2& mouseDelta = Input::GetMouseDelta();
+		m_pitch -= mouseDelta.y * lookSensitivityMouse;
+		m_yaw -= mouseDelta.x * lookSensitivityMouse;
+	}
+
+
+	// Lock pitch to the range -89 - 89
+	if ( m_pitch >= 89 ) {
+		m_pitch = 89;
+	}
+	else if ( m_pitch <= -89 ) {
+		m_pitch = -89;
+	}
+
+	// Lock yaw to the range 0 - 360
+	if ( m_yaw >= 360 ) {
+		m_yaw -= 360;
+	}
+	else if ( m_yaw <= 0 ) {
+		m_yaw += 360;
+	}
+
+	glm::vec3 forwards(
+		std::cos(glm::radians(m_pitch)) * std::cos(glm::radians(m_yaw)),
+		std::sin(glm::radians(m_pitch)),
+		std::cos(glm::radians(m_pitch)) * std::sin(glm::radians(m_yaw))
+	);
+	forwards = glm::normalize(forwards);
+
+	glm::vec3 forward = m_cam->getCameraDirection();
+	float totM = forwardM + backM + rightM + leftM + upM + downM;
+	//Logger::Warning(std::to_string(totM));
+	if ( totM != 0.f ) {
+		glm::vec3 dir = (forward * forwardM) - (forward * backM) + (right * rightM) - (right * leftM)
+			/*Only for flying*/ + ( m_cam->getCameraUp() * upM ) - ( m_cam->getCameraUp() * downM );
+
+		playerMovComp->setDirection(glm::normalize(dir));
```

Comment:
Fix normalization of 0-vector. Happens when two opposite directions are being pressed.